source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,188 @@ unittest
+            Ledger ledger, TaskManager taskman)
+        {
+            this.ledger = ledger;
+            super(network, key_pair, ledger, taskman);
+        }
+
+        ///
+        public override bool isNominating () @safe @nogc nothrow
+        {
+            return this.is_nominating;
+        }
+
+        ///
+        public override void tryNominate () @safe
+        {
+            if (this.ledger.getBlockHeight < 1)
+            {
+                super.tryNominate();
+            }
+        }
+    }
+
+    static class OneTimeValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new OneTimeNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 0)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                auto api = RemoteAPI!TestAPI.spawn!OneTimeValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+        }
+    }
+
+    auto network = makeTestNetwork!BadAPIManager(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto bad_validator = nodes[0];
+    auto validator = nodes[1];
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = validator.getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Make a bloack using Genesis
+    Transaction[] txs;
+    txs = genesisSpendable().take(Block.TxsInBlock).enumerate()
+        .map!(en => en.value.refund(WK.Keys.A.address).sign())
+        .array();
+    txs.each!(tx => validator.putTransaction(tx));
+
+    // Ensure everyone is at the same level
+    ensureConsistency(nodes, 1);
+
+    // Make a block in the `bad_validator`
+    txs = txs
+        .map!(ptx => TxBuilder(ptx).refund(WK.Keys.B.address).sign())
+        .array();
+    txs.each!(tx => bad_validator.putTransaction(tx));
+
```

Comment:
You need to add a Thread.sleep call here. Otherwise the `ensureConsistency()` check can be too fast, before nomination is even attempted.

Reply:
The function has already 2 seconds to wait for the result. Isn't that time enough?

Reply:
Are you satisfied with the new code @AndrejMitrovic ?

Reply:
Yes that's better.

Reply:
Just to make things clearer for everyone, what I was worried about was:

- checkConsistency(1)
- putTransactions...
- checkConsistency(1)

The last check could be too early because calling putTransaction will start an asynchronous nomination round (remember that tx's are gossiped to all nodes, they will start nominating too).

But Mathias fixed this by having only one checkConsistency() call. Consensus should still be reached even if one node nominates a bad enrollment - because the other nodes will select a valid set of transactions and enrollments.